File: 121. Best Time to Buy and Sell Stock.cpp
// Time: O(n), space: O(1)
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int buy = INT_MAX;
        int profit = 0;
        for (int price : prices) {
            buy = min(buy, price);
            profit = max(profit, price - buy);
        }
        
        return profit;
    }
};

class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int maxprofit = 0;
        int current = 0;
        for (int i = 1; i < prices.size(); ++i) {
            current = max(0, current + prices[i] - prices[i - 1]);
            maxprofit = max(maxprofit, current);
        }

        return maxprofit;
    }
};

File: 121. Best Time to Buy and Sell Stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        minprice = float('inf')
        profit = 0
        for p in prices:
            minprice = min(minprice, p)
            profit = max(profit, p - minprice)
        return profit

class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        if (not prices): return 0
        pro = maxpro = 0
        for i in range(1, len(prices)):
            pro += prices[i] - prices[i - 1]
            pro = max(pro, 0)
            maxpro = max(maxpro, pro)
        return maxpro
